velox/common/caching/SsdFile.cpp

```
@@ -0,0 +1,412 @@
+  stats_.bytesRead += regionBytes;
+  readFile_->preadv(offset, buffers);
+}
+
+std::pair<uint64_t, int32_t> SsdFile::getSpace(
+    const std::vector<CachePin>& pins,
+    int32_t begin) {
+  std::lock_guard<std::mutex> l(mutex_);
+  for (;;) {
+    if (writableRegions_.empty()) {
+      if (!growOrEvictLocked()) {
+        return {0, 0};
+      }
+    }
+    auto region = writableRegions_[0];
+    auto offset = regionSize_[region];
+    auto available = kRegionSize - regionSize_[region];
+    int64_t toWrite = 0;
+    for (; begin < pins.size(); ++begin) {
+      auto entry = pins[begin].entry();
+      if (entry->size() > available) {
+        break;
+      }
+      available -= entry->size();
+      toWrite += entry->size();
+    }
+    if (toWrite) {
+      // At least some pins got space from this region. If the region is is full
+      // the next call will get space from another region.
+      regionSize_[region] += toWrite;
+      return {region * kRegionSize + offset, toWrite};
+    }
+    // A region has been filled and transits from writable to
+    // evictable. Set its score to be at least the best score +
+    // kRegionSize so that it gets time to live. Otherwise it has had
+    // the least time to get hits and would be the first evicted.
+    uint64_t best = 0;
+    for (auto& score : regionScore_) {
+      best = std::max<uint64_t>(best, score);
+    }
+    regionScore_[region] = std::max(regionScore_[region], best + kRegionSize);
+    writableRegions_.erase(writableRegions_.begin());
+  }
+}
+
+bool SsdFile::growOrEvictLocked() {
+  if (numRegions_ < maxRegions_) {
+    auto newSize = (numRegions_ + 1) * kRegionSize;
+    auto rc = ftruncate(fd_, newSize);
+    if (rc >= 0) {
+      fileSize_ = newSize;
+      writableRegions_.push_back(numRegions_);
+      regionSize_[numRegions_] = 0;
+      ++numRegions_;
+      return true;
+    } else {
+      LOG(ERROR) << "Failed to grow cache file " << filename_ << " to "
```

Why is it OK to ignore this error?